### PR TITLE
Rnn language translation

### PR DESCRIPTION
**Pull request by ehoelzl**
This PR adds the GNMT model as well as the WMT14 dataset for Language translation task.

The model was taken directly from NVIDIA's [implementation](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Translation/GNMT), as well as all other aspects of training.

The original code was made for multi-GPU training on a single node, and uses `NCCL` backend. Additionally, it uses mixed precision training using [apex](https://nvidia.github.io/apex/amp.html).

The code uses `horovod` with MPI, as float16 tensor reduction is not supported yet

### CODE REVIEW

File: mlbench_core/cli/cli.py [hunk 1]
```diff
@@ -567,6 +567,7 @@ def create_gcloud(
                 "source": {
                     "type": "git",
                     "location": "https://github.com/mlbench/mlbench-helm",
+                    "reference": "RNN-Language-translation",
```
Comment (ehoelzl) [inline mlbench_core/cli/cli.py]: this change needs to be removed before merging

File: mlbench_core/evaluation/pytorch/inference/beam_search.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
+import torch
+
+from mlbench_core.dataset.translation.pytorch.config import BOS, EOS
+
+
+class SequenceGenerator:
```
Comment (IamTao) [inline mlbench_core/evaluation/pytorch/inference/beam_search.py]: so you implement it yourself or just borrow from another library? a bit worried about the efficiency of the beam search proposed here.

Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/inference/beam_search.py:6]: This has been taken directly from NVIDIA code. It generally uses the `beam_search` instead of `greedy_search`.

File: mlbench_core/optim/pytorch/optim.py [hunk 3]
```diff
@@ -462,6 +480,109 @@ def step(self, closure=None):
...
+            gradients
+        """
+        self.model = model
+        self.grad_clip = grad_clip
+        self.optimizer = optimizer
+        loss_scaler = amp._amp_state.loss_scalers[0]
+        loss_scaler._loss_scale = loss_scale
+        loss_scaler._scale_seq_len = dls_upscale_interval
+        self.agg = AllReduceAggregation(
+            world_size=world_size, cast_in=torch.float32, cast_out=torch.float16
```
Comment (IamTao) [inline mlbench_core/optim/pytorch/optim.py:563]: BTW, I found [horovod](https://github.com/horovod/horovod) can support FP16 gradient reduction via their modified MPI. maybe it is worth to have a [look](https://github.com/horovod/horovod/blob/4371007e154b7882062cab8696b26d520252971b/test/test_torch.py#L179).

Comment (ehoelzl) [reply mlbench_core/optim/pytorch/optim.py:563]: Okay, I will check this.

File: mlbench_core/evaluation/pytorch/inference/beam_search.py (hunk 2 above)
Comment (martinjaggi) [reply mlbench_core/evaluation/pytorch/inference/beam_search.py:6]: good to mark all existing code used clearly as such, then it also makes it easier in the future to upgrade and detangle things

**Comment by ehoelzl** [comment]
> if we make this communication variant permanent, can we make sure the other benchmarks still run with and without horovod (that it is supported as an additional but not mandatory backend). 
also now it's not so clearly documented how this function depends on what backend the user has chosen (it basically overrides that maybe, which is ok but should be documented)

@martinjaggi 
Horovod uses the backend that is set by `torch.distributed`, so in that case it uses MPI, and works with NCCL and GLOO too. The only reason we are using it here is because it allows us to reduce `float16` tensors, which `torch.distributed` doesn't.

File: mlbench_core/evaluation/pytorch/inference/beam_search.py (hunk 2 above)
Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/inference/beam_search.py:7]: I added a comment with the link to the repo where it was taken from.

File: mlbench_core/optim/pytorch/fp_optimizers.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+        # Aggregate gradients
+        self.agg(self.fp16_model, self.agg_mode)
+        # Cast fp16 params to fp32 for optimizer
+        self.fp16_to_fp32_flat_grad(self.fp32_params, self.fp16_model)
+
+        if scaling_factor != 1.0:
+            self.fp32_params.grad.data /= scaling_factor
+        start = datetime.now()
+        torch.cuda.synchronize()
```
Comment (ehoelzl) [inline mlbench_core/optim/pytorch/fp_optimizers.py:177]: @IamTao Here is the line where it blocks. At every call for `step()`, all processes stall for between 3 and 5 seconds here.
I have been looking into that for a few days now. Not sure how to solve it. Maybe it is due to GPUs not being connected through NVLink ?

**Comment by ehoelzl** [comment]
@martinjaggi Some interesting findings with this benchmark:
This model is trained using mixed precision training, meaning gradients are in `float16` but optimizers work on `float32` weights. This makes backpropagation twice as fast, and theoretically, makes communication of `float16` gradients faster.

First issue I encountered was that MPI does not support reduction of `float16` with regular `dist.all_reduce`. To get around that, I used [`horovod`](https://github.com/horovod/horovod), which allows reduction of `float16` tensors through MPI. 
After that I tried running on NCCL, expecting better results. However, to my surprise, NCCL turned out to be 2-3 times slower than MPI. Basically, after each `all_reduce` call, the training stalled for around 5 seconds. 

I looked into that for a few days, thinking it was the data loader or GPU memory issues. I finally tried running `NCCL + Horovod` instead of `NCCL + pytorch.distributed`, and this actually solved my problem, making NCCL effectively faster than MPI.

After some debugging and some help from @IamTao , I figured out that NCCL operations are "synchronous", but really are async since they are put in the GPU queue. I don't really know how horovod works around that, but the stalling does not happen anymore, and convergence is identical.

Hence, I thought it would be interesting to study the difference of using `NCCL + pytorch.distributed` vs `NCCL + horovod` for reduction of large `float16` tensors (the tensor is around 200MB), and investigate why this is.

**Comment by martinjaggi** [comment]
agreed, very interesting, and would be nice to write down the results on this (blog post or report). can you break it down a bit to have the NLP benchmark itself, then compare fp16 to fp32, and finally the horovod issue (the latter two items are also interesting for the vision task, right?)

**Comment by ehoelzl** [comment]
> agreed, very interesting, and would be nice to write down the results on this (blog post or report). can you break it down a bit to have the NLP benchmark itself, then compare fp16 to fp32, and finally the horovod issue (the latter two items are also interesting for the vision task, right?)

Sure. It can be broken down pretty easily. For the benchmark task, i'll stick to automatic mixed precision (using `apex.amp`) using `horovod` as it is the fastest training.

I have implemented 3 optimizers for this task:
- Float 32 optimizer, using gradient clipping
- Manual mixed precision: everything is coded by hand using no libraries
- Automatic Mixed Precision using AMP.

Then we can also use horovod or not and study the differences for each Backend.

Can this PR be merged so that all new code is integrated into `develop` ?

**Comment by martinjaggi** [comment]
LGTM!

have you checked that the 3 optimizers work fine with the vision tasks too (or can do later after this is merged)

**Comment by ehoelzl** [comment]
> LGTM!
> 
> have you checked that the 3 optimizers work fine with the vision tasks too (or can do later after this is merged)

No I haven't tested with Vision tasks, but I'm not sure if they use mixed precision training. And yes we can test once it is merged

File: mlbench_core/evaluation/pytorch/inference/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from .translator import Translator
```
Comment (Panaetius) [inline mlbench_core/evaluation/pytorch/inference/__init__.py:1]: Maybe the classes in the `inference` namespace would better fit in the `mlbench_core.utils` namespace or alongside the model. the `Evaluation` namespace is meant for losses and measures like accuracy, f1 score etc.

File: mlbench_core/lr_scheduler/pytorch/lr.py [hunk 6]
```diff
@@ -342,3 +342,87 @@ def __init__(self, optimizer, alpha):
 
     def f(self, iteration):
         return 1.0 / math.sqrt(max(1, iteration))
+
+
+def perhaps_convert_float(param, total):
```
Comment (Panaetius) [inline mlbench_core/lr_scheduler/pytorch/lr.py:347]: Use a more descriptive name, like `convert_relative_stepsize` or something along those lines so the intention is more clear.

File: mlbench_core/models/pytorch/gnmt/attention.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+from torch.nn.parameter import Parameter
+
+
+class BahdanauAttention(nn.Module):
+    """
+    Bahdanau Attention (https://arxiv.org/abs/1409.0473)
+    Implementation is very similar to tf.contrib.seq2seq.BahdanauAttention
+
+    Args:
+        query_size: feature dimension for query
```
Comment (Panaetius) [inline mlbench_core/models/pytorch/gnmt/attention.py:15]: `Args:` in docstrings should specify their type in brackets like `count (int): The count for something, Default: '0'`, this is missing in this file

File: mlbench_core/controlflow/pytorch/gnmt/controlflow.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    Args:
+        batch_idx (int): The id of the current batch
+        loss (float): The loss of the batch
+        tracker (`obj`:mlbench_core.utils.Tracker): Tracker object to use.
+        num_batches_per_device_train (int): Number of batches per train epoch
+    """
+    progress = batch_idx / num_batches_per_device_train
+    progress += tracker.current_epoch
+
+    log_to_api = (
```
[the remark this answers is not in the document]
Comment (ehoelzl) [reply mlbench_core/controlflow/pytorch/gnmt/controlflow.py:27]: The problem is that training accuracy slows down training a lot (because the translator needs to be run and it adds significant lag)

File: mlbench_core/controlflow/pytorch/gnmt/controlflow.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+        """Sets the model and tracker in evaluation mode"""
+        self.model.eval()
+
+        # Set tracker in validation mode
+        if self.tracker:
+            self.tracker.validation()
+            self.tracker.validation_start()
+
+    def train_round(
+        self, train_loader, val_loader, bleu_score=False, validate_every=None
```
[the remark this answers is not in the document]
Comment (ehoelzl) [reply mlbench_core/controlflow/pytorch/gnmt/controlflow.py:237]: So basically completely remove the class and keep it functional ? I could do that, but i would need to move all the functions related to GNMT specifically (like `compute_model_output` and even `optimize`). Or I could simply change the functions `train_round` and `validation_round` and pass the trainer object to it.

We could create functions like `train_round` and `validation_round`, which implement a certain logic that isn't specific to any model, and make them take a instance of an abstract class we define (e.g. `ModelTrainer`) which should define some functions that will be called by `train_round` and `validation_round`.

Let me know what you think

File: mlbench_core/dataset/translation/pytorch/config.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
+"""Configuration for WMT14 dataset"""
```
[the remark this answers is not in the document]
Comment (ehoelzl) [reply mlbench_core/dataset/translation/pytorch/config.py:1]: so the structure would be `mlbench_core/dataset/nlp/pytorch/translation/` ? As there are multiple files

[the remark this answers is not in the document]
Comment (Panaetius) [reply mlbench_core/dataset/translation/pytorch/config.py:1]: yes I think so. And maybe add the main class in `mlbench_core/dataset/nlp/pytorch/` for ease of access
